src/main/java/org/pac4j/vertx/handler/impl/CallbackHandler.java

```diff
@@ -34,75 +40,67 @@
 
-    wrapper.authenticate(routingContext, clientName, sessionAttributes, handler);
+  protected String defaultUrl = Pac4jConstants.DEFAULT_URL_VALUE;
 
+  public CallbackHandler(final Vertx vertx,
+                           final Config config) {
+      this.vertx = vertx;
+      this.config = config;
   }
 
   @Override
-  protected void retrieveUserProfile(final RoutingContext routingContext, final Pac4jSessionAttributes sessionAttributes,
-                                     final Handler<Pac4jAuthenticationResponse> handler) {
-    authenticate(routingContext, sessionAttributes, handler);
-  }
-
-  @Override
-    protected void authenticationSuccess(final RoutingContext routingContext, final Pac4jSessionAttributes sessionAttributes) {
-
-        redirectToTarget(routingContext, sessionAttributes);
-    }
-
-    protected void authenticationFailure(final RoutingContext routingContext, final Pac4jSessionAttributes sessionAttributes) {
-
-        redirectToTarget(routingContext, sessionAttributes);
-    }
-
-    private void redirectToTarget(final RoutingContext routingContext, final Pac4jSessionAttributes sessionAttributes) {
-
-      String requestedUrl = retrieveOriginalUrl(sessionAttributes);
-      saveUrl(null, sessionAttributes);
-      final String redirectUrl = defaultUrl(requestedUrl, BaseConfig.getDefaultSuccessUrl());
-      pac4jAuthProvider.saveSessionAttributes(routingContext, sessionAttributes);
-      HttpResponseHelper.redirect(routingContext.request(), redirectUrl);
-
+  public void handle(RoutingContext event) {
+
+    // Can we complete the authentication process here?
+    final VertxWebContext webContext = new VertxWebContext(event);
+    final ProfileManager profileManager = new ProfileManager(webContext);
+    final Client client = config.getClients().findClient(webContext);
+
+    CommonHelper.assertNotNull("client", client);
+
+    final Credentials credentials;
+    try {
+      credentials = client.getCredentials(webContext);
+    } catch (final RequiresHttpAction e) {
+      LOG.warn("Requires http action: " + e.getCode());
+      httpActionHandler.handle(e.getCode(), webContext);
+      return;
     }
-
-  /**
-   * This method returns the default url from a specified url compared with a default url.
-   *
-   * @param url a specific url
-   * @param defaultUrl the default url
-   * @return the default url
-   */
-  public static String defaultUrl(final String url, final String defaultUrl) {
-    String redirectUrl = defaultUrl;
-    if (url != null && !"".equals(url)) {
-      redirectUrl = url;
-    }
-    logger.debug("defaultUrl : {}", redirectUrl);
-    return redirectUrl;
+    vertx.<CommonProfile>executeBlocking(future -> {
+      final CommonProfile profile = (CommonProfile) client.getUserProfile(credentials, webContext);
+      future.complete(profile);
+    }, result -> {
+      if (result.succeeded()) {
+        Optional.ofNullable(result.result()).ifPresent(commonProfile -> {
```

Comment:
I have turned that into a separate method as well. When I'll be done on my side, you'll take at all the changes and report back them if possible.